code/modules/flufftext/Hallucination.dm

```diff
@@ -1191,21 +1359,51 @@ GLOBAL_LIST_INIT(hallucination_list, list(
 		target.client.images += fire_overlay
 	to_chat(target, "<span class='userdanger'>You're set on fire!</span>")
 	target.throw_alert("fire", /atom/movable/screen/alert/fire, override = TRUE)
-	sleep(20)
-	for(var/i in 1 to 3)
-		if(target.fire_stacks <= 0)
-			clear_fire()
-			return
-		stage++
-		update_temp()
-		sleep(30)
-	for(var/i in 1 to rand(5, 10))
-		if(target.fire_stacks <= 0)
-			clear_fire()
-			return
-		target.adjustStaminaLoss(15)
-		sleep(20)
-	clear_fire()
+	times_to_lower_stamina = rand(5, 10)
+	addtimer(CALLBACK(src, .proc/start_expanding), 20)
+
+/datum/hallucination/fire/Destroy()
+	. = ..()
+	STOP_PROCESSING(SSfastprocess, src)
+
+/datum/hallucination/fire/proc/start_expanding()
+	if (isnull(target))
+		qdel(src)
+		return
+	START_PROCESSING(SSfastprocess, src)
+
+/datum/hallucination/fire/process(delta_time)
+	if (isnull(target))
+		qdel(src)
+		return
+
+	if(target.fire_stacks <= 0)
+		clear_fire()
+
+	time_spent += delta_time
+
+	if (fire_clearing)
+		next_action -= delta_time
+		if (next_action < 0)
+			stage -= 1
+			update_temp()
+			next_action += 3
+	else if (increasing_stages)
+		var/new_stage = min(round(time_spent / RAISE_FIRE_TIME), RAISE_FIRE_COUNT)
+		if (stage != new_stage)
+			stage = new_stage
+			update_temp()
+
+			if (stage == RAISE_FIRE_COUNT)
+				increasing_stages = FALSE
+	else if (times_to_lower_stamina)
+		next_action -= delta_time
+		if (next_action < 0)
+			target.adjustStaminaLoss(15)
+			next_action += 2
```

Comment:
I assume this was necessary, but this is painful

Reply:
all this weak for the least fun hallucination too